File: java/collatz-conjecture/src/main/java/CollatzCalculator.java
class CollatzCalculator {

    int computeStepCount(int start) {
        if (start <= 0) {
            throw new IllegalArgumentException("Only natural numbers are allowed");
        }

        int step = 0;
        int i = start;
        while (i > 1) {
            step++;

            if (i % 2 == 0) {
                i = i / 2;
            } else {
                i = (i * 3) + 1;
            }
        }

        return step;
    }

}

File: kotlin/collatz-conjecture/src/main/kotlin/CollatzCalculator.kt
object CollatzCalculator {
    fun computeStepCount(start: Int): Int {
        require(start > 0) { "Only positive integers allowed" }

        return countSteps(start, 0)
    }

    private fun countSteps(n: Int, stepNumber: Int): Int {
        when {
            n == 1 -> return stepNumber
            n.IsEven() -> return countSteps(n / 2, stepNumber + 1)
            else -> return countSteps(n * 3 + 1, stepNumber + 1)
        }
    }
}

private fun Int.IsEven() = this % 2 == 0
